File: frontend/demo/component/timepicker/time-picker-min-max.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/time-picker';
import type { TimePickerChangeEvent } from '@vaadin/time-picker';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('time-picker-min-max')
export class Example extends LitElement {
  @state()
  protected errorMessage = '';

  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-time-picker
        label="Appointment time"
        helper-text="Open 8:00-16:00"
        value="08:30"
        min="08:00"
        max="16:00"
        .step="${60 * 30}"
        error-message="${this.errorMessage}"
        @change="${(event: TimePickerChangeEvent) => {
          const { min, max, value } = event.target;
          if (value < min) {
            this.errorMessage = 'Too early, choose another time';
          } else if (value > max) {
            this.errorMessage = 'Too late, choose another time';
          } else {
            this.errorMessage = '';
          }
        }}"
      ></vaadin-time-picker>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/timepicker/TimePickerMinMax.java
package com.vaadin.demo.component.timepicker;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.timepicker.TimePicker;
import com.vaadin.flow.router.Route;

import java.time.Duration;
import java.time.LocalTime;

import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("time-picker-min-max")
public class TimePickerMinMax extends Div {

    public TimePickerMinMax() {
        // tag::snippet[]
        TimePicker timePicker = new TimePicker();
        timePicker.setLabel("Appointment time");
        timePicker.setHelperText("Open 8:00-16:00");
        timePicker.setStep(Duration.ofMinutes(30));
        timePicker.setValue(LocalTime.of(8, 30));
        timePicker.setMin(LocalTime.of(8, 0));
        timePicker.setMax(LocalTime.of(16, 0));
        timePicker.addValueChangeListener(event -> {
            LocalTime value = event.getValue();
            String errorMessage = null;
            if (value != null) {
                if (value.compareTo(timePicker.getMin()) < 0) {
                    errorMessage = "Too early, choose another time";
                } else if (value.compareTo(timePicker.getMax()) > 0) {
                    errorMessage = "Too late, choose another time";
                }
            }
            timePicker.setErrorMessage(errorMessage);
        });
        add(timePicker);
        // end::snippet[]
    }

    public static class Exporter extends DemoExporter<TimePickerMinMax> { // hidden-source-line
    } // hidden-source-line
}
